### PR TITLE
Lock all packages in workspace

### LINKED ISSUE #3983: Locking in a workspace should lock the entire workspace
If you run `uv lock` in a workspace, we should lock the entire workspace, creating a single shared lockfile.

### PR DESCRIPTION
**Pull request by konstin**
When creating a lockfile, lock the combined dependencies for all packages in a workspace. This make the lockfile independent of where you are in the workspace.

Fixes #3983

### CODE REVIEW

File: crates/uv-distribution/src/workspace.rs [hunk 1]
```diff
@@ -105,7 +105,11 @@ impl Workspace {
             {
                 workspace
             } else {
-                return Err(WorkspaceError::MissingWorkspace(project_path));
+                (
+                    project_path.clone(),
+                    ToolUvWorkspace::default(),
+                    PyProjectToml::default(),
```
Comment (charliermarsh) [inline crates/uv-distribution/src/workspace.rs:111]: What's the motivation for this change? Why doesn't it return `pyproject_toml`?

File: crates/uv-distribution/src/workspace.rs [hunk 2]
```diff
@@ -153,6 +157,44 @@ impl Workspace {
...
+                let project = member.pyproject_toml.project.as_ref()?;
+                // Extract the extras available in the project.
+                let extras = project
+                    .optional_dependencies
+                    .as_ref()
+                    .map(|optional_dependencies| {
+                        // It's a `BTreeMap` so the keys are sorted.
+                        optional_dependencies.keys().cloned().collect::<Vec<_>>()
+                    })
+                    .unwrap_or_default();
```
Comment (charliermarsh) [inline crates/uv-distribution/src/workspace.rs:174]: We already have this on `ProjectWorkspace`. Is there any way to get it from there...?

**Comment by charliermarsh** [comment]
I think we need to find some solution to `Requires-Python` now. What is the `Requires-Python` of the workspace?

File: crates/uv-distribution/src/workspace.rs (hunk 1 above)
Comment (konstin) [reply crates/uv-distribution/src/workspace.rs]: An oversight, fixed now

**Comment by konstin** [comment]
I can either be the intersection or the union of the python requirement of the constituent packages. We could also force them to identical but that seems to strict.

Let's say A has python >= 3.9 and B python >=3.10.
Union: We require the workspace to be compatible with >= 3.9. This means we can install A alone on 3.9, but requires all dependencies of B to be compatible with 3.9 too (unless you go for marker tricks)
Intersection: We require the workspace to be compatible with >=3.10. This means you can't test with A on 3.9, but B only gets resolved on >=3.10 only.

From that, the union seems slightly better.

**Comment by konstin** [comment]
Created https://github.com/astral-sh/uv/pull/4041 that solves this properly

File: crates/uv-distribution/src/workspace.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv-distribution/src/workspace.rs]: But what does it mean? What case is this?

Comment (konstin) [reply crates/uv-distribution/src/workspace.rs]: Oh you meant the branch not the line, this is for supporting implicit workspace roots, i've added a comment now.

**Comment by charliermarsh** [comment]
I will review and merge today when I'm out of meetings.

**Review by charliermarsh — APPROVED** [review]
